Improved support for Dancer and Dancer2

ulizama opened the pull request:
#84 This pull requests does some improvements on the Dancer/Dancer2 support.

- Added some missing keywords like `ajax`, `patch`, `del`
- Added support for the `hook` syntax on Dancer
- Added support for regexp and wildcard matches
- Correctly manage multi request routes: `any ["get", 'post'] => "/route/any" => sub {`
- Added the display of the type of request on the symbol name: `/my/route` => `get /my/route` to make it easier to differentiate between routes using the same name


![Screenshot 2023-08-29 at 12 53 05](https://github.com/bscan/PerlNavigator/assets/581703/8a1b2952-b106-4194-baae-26b5e79fae60)

bscan commented on syntaxes/dancer.json:
```diff
@@ -15,7 +15,7 @@
         "dancer": {
             "patterns": [
                 {
-                    "match": "^\\s*\\b(hook)\\s+(\\w+)\\s+(?==>\\s*sub\\b)",
```
This would no longer recognize:
```perl
hook before => sub {
```
All the examples I saw in the Dancer manual used `=>` in them. Although if it's optional, perhaps you could match the regex from pltags? `(?:=>)?`

bscan commented on server/src/perl/lib_bs22/pltags.pm:
```diff
@@ -324,14 +325,32 @@ sub build_pltags {
-            my $route = $2;
+            $stmt=~/^(?:any|before\_route)\s+\[([^\]]+)\]\s+(?:=>\h*)?(['"])([^\2]+)\2\s*=>\s*sub/) { # Multiple request routing paths
+            my $requests = $1;
+            if ($requests) {
```
I believe `$requests` will always be truthy because of the + in `([^\]]+)`

ulizama replied:
Agreed, removed the if.

bscan reviewed (changes requested):
Looks great. Thanks for following up on this. My primary feedback is around the regex in `CleanForBalanced` since it gets applied to all code in all subroutines. Would be safer to keep that in the Dancer regex routing section. Otherwise, some comments inline. Thanks!

bscan commented:
Also, in your screenshot, it doesn't look like `get/any/post/hook` keywords were syntax highlighted. Does the additional syntax highlighting work for you? If you have other syntax highlighting extensions enabled (e.g. `Better Perl Syntax`) they can interfere with the Navigator's highlighting.

ulizama commented on server/src/perl/lib_bs22/pltags.pm:
```diff
@@ -109,6 +109,7 @@ sub CleanForBalanced {
     # $input =~ s@\s//(\n|\s)@ ||$1@g; 
     $input =~ s/[^\x00-\x7F]/ /g;
     $input =~ s/->@\*/ /g;
+    $input =~ s/qr\{[^\}]+\}/ /g;
```
Before doing the qr replacement I'm checking if this is the start of a Dancer method.

ulizama commented:
Updated the code with the suggested fixes.

I also updated the screenshot with the Better Perl Syntax disabled to show the correct highlights from the plugin. All Dancer keywords are being correctly highlighted.

bscan commented:
Thanks for the commit and I appreciate your patience on this one. While I was merging your code, it made me realize I had a few long-standing issues in there. For example, pltags was double looping over all code in `CleanForBalanced` which was slower than needed.  By preprocessing the code and refactoring into a class, the tagger is now better suited for extensions like this one. For example, the Dancer2 code is now it's own function: https://github.com/bscan/PerlNavigator/blob/a270c5e5eab1f1c4091eee30163a1300dc6c2a11/server/src/perl/lib_bs22/pltags.pm#L376 

I like the idea of supporting more frameworks and your Dancer2 example is great code for others to follow. Thanks again for the ticket and the pull request. I will do some more testing before releasing to the marketplace, but let me know if I broke anything during the merge. Thanks!
